File: src/calendar.ts
import pLimit from "p-limit";
import { createError } from "@fastify/error";
import dayjs from "dayjs";

import type { Collection, Episode, Paged, Subject } from "./bangumi";
import type { Cache } from "./cache";
import { get } from "./request";
import { notNull, unix, uuidByString } from "./util";
import { logger } from "./logger";

const NotFoundError = createError("NOT_FOUND", "%s", 404);

const limit = pLimit(20);

export async function buildICS(username: string, cache: Cache): Promise<string> {
  logger.info(`fetching collection of user ${username}`);
  let collections: Array<Collection> = await fetchAllUserCollection(username);

  const subjects: SlimSubject[] = (
    await Promise.all(collections.map((s) => limit(() => getSubjectInfo(s.subject_id, cache))))
  )
    .filter(notNull)
    .filter((s) => s.future_episodes.length !== 0);

  return renderICS(subjects);
}

async function fetchAllUserCollection(username: string, pageSize: number = 50): Promise<Array<Collection>> {
  const data: Array<Collection> = [];

  for (const collectionType of [1, 3]) {
    let offset: number = 0;
    let body: Paged<Collection>;
    do {
      const res = await get(`v0/users/${username}/collections`, {
        query: {
          type: collectionType,
          offset,
          limit: pageSize,
        },
      });

      if (!res.ok) {
        if (res.code === 404) {
          throw new NotFoundError(`user ${username} not found`);
        }

        throw new Error(`Unexpected HTTP ERROR ${res.body}`);
      }

      body = JSON.parse(res.body) as Paged<Collection>;

      data.push(
        ...body.data.filter((c) => c.subject_type === SubjectTypeAnime || c.subject_type === SubjectTypeEpisode),
      );

      offset += pageSize;
    } while (offset < body.total);
  }

  return data;
}

interface SlimSubject {
  name: string;
  id: number;
  future_episodes: Array<ParsedEpisode>;
}

async function getSubjectInfo(subjectID: number, cache: Cache): Promise<SlimSubject | null> {
  const cacheKey = `subject-v3-${subjectID}`;

  const cached = await cache.get<SlimSubject>(cacheKey);
  if (cached !== undefined) {
    return cached;
  }

  let data: SlimSubject;
  let total_episode = 0;

  logger.info(`fetching subject ${subjectID}`);
  const { body, code } = await get(`v0/subjects/${subjectID}`);
  if (code === 404) {
    await cache.set(cacheKey, null, 60 * 60 * 24);
    return null;
  }
  const s = JSON.parse(body) as Subject;
  total_episode = s.total_episodes;
  data = {
    id: s.id,
    name: s.name_cn || s.name,
    future_episodes: [],
  };

  if (total_episode) {
    const all_episodes = await fetchAllEpisode(subjectID);

    const today = unix() - 24 * 60 * 60 * 3;

    const future_episodes = all_episodes.filter((episode) => {
      const ts = dayjs(new Date(episode.air_date[0], episode.air_date[1] - 1, episode.air_date[2]));
      return ts.unix() > today;
    });

    if (all_episodes.length !== 0 && future_episodes.length === 0 && all_episodes.length <= 200) {
      // no future episodes, just cache it longer than normal episode
      await cache.set(cacheKey, data, 60 * 60 * 24 * 7);
      return data;
    }

    data.future_episodes.push(...future_episodes);
  }

  await cache.set(cacheKey, data, 60 * 60 * 24 * 3);

  return data;
}

interface ParsedEpisode {
  id: number;
  sort: number;
  name: string;
  air_date: readonly [number, number, number];
  duration: string;
}

async function fetchAllEpisode(subjectID: number): Promise<Array<ParsedEpisode>> {
  const res = await _fetchAllEpisode(subjectID);

  return res
    .map((episode): ParsedEpisode | null => {
      const date: number[] = episode.airdate
        .split("-")
        .map((x) => parseInt(x, 10))
        .filter((x) => !isNaN(x))
        .filter((x) => x !== null);
      if (date.length != 3) {
        return null;
      }

      const [year, month, day] = date;

      if (year === undefined || month === undefined || day === undefined) {
        return null;
      }

      return {
        id: episode.id,
        sort: episode.sort,
        name: episode.name_cn || episode.name,
        air_date: [year, month, day],
        duration: episode.duration,
      };
    })
    .filter(notNull);
}

async function _fetchAllEpisode(subjectID: number, pageSize: number = 200): Promise<Array<Episode>> {
  const data: Array<Episode> = [];
  let offset: number = 0;
  let res: Paged<Episode>;

  do {
    const { code, body } = await get("v0/episodes", {
      query: {
        subject_id: subjectID,
        offset,
        limit: pageSize,
      },
    });

    if (code !== 200) {
      throw new Error(`unexpected error ${code} ${JSON.stringify(body)}`);
    }

    res = JSON.parse(body) as Paged<Episode>;

    data.push(...res.data);

    offset += pageSize;
  } while (offset < res.total);

  return data;
}

const SubjectTypeAnime = 2;
const SubjectTypeEpisode = 6;

function renderICS(subjects: SlimSubject[]): string {
  const calendar = new ICalendar({
    name: "Bangumi Episode Air Calendar",
    "X-PUBLISHED-TTL": "P1D",
  });
  const today = unix();

  for (const subject of subjects) {
    for (const episode of subject.future_episodes) {
      const date = episode.air_date;
      const ts = dayjs(new Date(date[0], date[1] - 1, date[2]));

      // only show episode in 30 days.
      if (ts.unix() > today + 30 * 24 * 60 * 60) {
        continue;
      }

      const end = new Date(date[0], date[1] - 1, date[2]);
      end.setDate(end.getDate() + 1);

      calendar.createEvent({
        subjectID: subject.id,
        episodeID: episode.id,
        start: formatDate(date),
        end: formatDate([end.getFullYear(), end.getMonth() + 1, end.getDate()]),
        summary: `${subject.name} ${episode.sort}`,
        description: episode.name || undefined,
        duration: episode.duration,
      });
    }
  }

  return calendar.toString();
}

interface Event {
  subjectID: number;
  episodeID: number;
  start: string;
  end: string;
  summary: string;
  description?: string;
  duration: string;
}

class ICalendar {
  private readonly name: string;
  private readonly now: Date;
  private readonly lines: string[];

  constructor({ name, ...meta }: { name: string; [keys: string]: string }) {
    this.name = name;
    this.now = new Date();
    this.lines = [
      "BEGIN:VCALENDAR",
      "VERSION:2.0",
      "PRODID:-//trim21//bangumi-icalendar//CN",
      `NAME:${this.name}`,
      `X-WR-CALNAME:${this.name}`,
    ];

    this.lines.push(...Object.entries(meta).map(([key, value]) => `${key}:${value}`));
  }

  createEvent(event: Event): void {
    this.lines.push(
      "BEGIN:VEVENT",
      `UID:${generateUID(`subject-${event.subjectID}-episode-${event.episodeID}`)}`,
      `DTSTAMP:${formatDateObject(this.now)}`,
      `DTSTART;VALUE=DATE:${event.start}`,
      `DTEND;VALUE=DATE:${event.end}`,
      `SUMMARY:${event.summary}`,
    );

    let description: string[] = [];

    if (event.description) {
      description.push(event.description);
    }

    if (event.duration) {
      description.push("时长：" + event.duration);
    }

    if (description.length) {
      this.lines.push(`DESCRIPTION:${description.join("\\n\\n")}`);
    }

    this.lines.push("END:VEVENT");
  }

  toString(): string {
    return this.lines.join("\n") + "\nEND:VCALENDAR";
  }
}

function formatDateObject(d: Date): string {
  return [
    d.getUTCFullYear(),
    pad(d.getUTCMonth() + 1),
    pad(d.getUTCDate()),
    "T",
    pad(d.getUTCHours()),
    pad(d.getUTCMinutes()),
    pad(d.getUTCSeconds()),
    "Z",
  ].join("");
}

function formatDate(d: readonly [number, number, number]): string {
  return d[0].toString().padStart(4) + pad(d[1]) + pad(d[2]);
}

function pad(n: number) {
  return n < 10 ? `0${n}` : `${n}`;
}

function generateUID(summary: string): string {
  return uuidByString(summary);
}

File: src/calendar.rs
use chrono::{Datelike, Timelike, Utc};
use uuid::Uuid;

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct SlimSubject {
    pub name: String,
    pub id: i64,
    pub future_episodes: Vec<ParsedEpisode>,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct ParsedEpisode {
    pub id: i64,
    pub sort: f64,
    pub name: String,
    pub air_date: [i32; 3],
    pub duration: String,
}

#[derive(Debug, Clone)]
struct Event {
    subject_id: i64,
    episode_id: i64,
    start: String,
    end: String,
    summary: String,
    description: String,
}

const NAMESPACE: &str = "ef2256c4-162e-446b-9ccf-81050809d0c9";

pub fn render_ics(subjects: &[SlimSubject]) -> String {
    let mut cal = ICalendar::new(
        "Bangumi Episode Air Calendar",
        [
            ("X-PUBLISHED-TTL", "PT8H"),
            ("REFRESH-INTERVAL;VALUE=DURATION", "PT8H"),
        ],
    );

    let now = Utc::now();
    let max_ts = now.timestamp() + 30 * 24 * 60 * 60;

    for subject in subjects {
        for ep in &subject.future_episodes {
            let start_date = chrono::NaiveDate::from_ymd_opt(
                ep.air_date[0],
                ep.air_date[1] as u32,
                ep.air_date[2] as u32,
            );
            let Some(start_date) = start_date else {
                continue;
            };

            let start_dt = start_date.and_hms_opt(0, 0, 0).unwrap();
            if start_dt.and_utc().timestamp() > max_ts {
                continue;
            }

            let end_date = start_date.succ_opt().unwrap_or(start_date);

            let mut description_parts = vec![format!("https://bgm.tv/ep/{}", ep.id)];
            if !ep.name.is_empty() {
                description_parts.push(ep.name.clone());
            }
            if !ep.duration.is_empty() {
                description_parts.push(format!("时长：{}", ep.duration));
            }
            let description = description_parts.join("\\n");

            cal.add_event(Event {
                subject_id: subject.id,
                episode_id: ep.id,
                start: format_date_array(ep.air_date),
                end: format_date(end_date),
                summary: format!("{} {}", subject.name, ep.sort),
                description,
            });
        }
    }

    cal.finish()
}

struct ICalendar {
    now: chrono::DateTime<Utc>,
    lines: Vec<String>,
}

impl ICalendar {
    fn new(name: &str, meta: impl IntoIterator<Item = (&'static str, &'static str)>) -> Self {
        let mut lines = vec![
            "BEGIN:VCALENDAR".to_string(),
            "VERSION:2.0".to_string(),
            "PRODID:-//trim21//bangumi-icalendar//CN".to_string(),
            format!("NAME:{}", name),
            format!("X-WR-CALNAME:{}", name),
        ];
        for (k, v) in meta {
            lines.push(format!("{}:{}", k, v));
        }
        Self {
            now: Utc::now(),
            lines,
        }
    }

    fn add_event(&mut self, e: Event) {
        self.lines.push("BEGIN:VEVENT".to_string());
        self.lines.push(format!(
            "UID:{}",
            generate_uid(&format!(
                "subject-{}-episode-{}",
                e.subject_id, e.episode_id
            ))
        ));
        self.lines
            .push(format!("DTSTAMP:{}", format_datetime(self.now)));
        self.lines.push(format!("DTSTART;VALUE=DATE:{}", e.start));
        self.lines.push(format!("DTEND;VALUE=DATE:{}", e.end));
        self.lines.push(format!("SUMMARY:{}", e.summary));
        if !e.description.is_empty() {
            self.lines.push(format!("DESCRIPTION:{}", e.description));
        }
        self.lines.push("END:VEVENT".to_string());
    }

    fn finish(self) -> String {
        // Go 版本：strings.Join(lines, "\n") + "\nEND:VCALENDAR"
        let mut out = self.lines.join("\n");
        out.push_str("\nEND:VCALENDAR");
        out
    }
}

fn format_date_array(d: [i32; 3]) -> String {
    format!("{:04}{:02}{:02}", d[0], d[1], d[2])
}

fn format_date(d: chrono::NaiveDate) -> String {
    format!("{:04}{:02}{:02}", d.year(), d.month(), d.day())
}

fn format_datetime(d: chrono::DateTime<Utc>) -> String {
    format!(
        "{:04}{:02}{:02}T{:02}{:02}{:02}Z",
        d.year(),
        d.month(),
        d.day(),
        d.hour(),
        d.minute(),
        d.second()
    )
}

fn generate_uid(summary: &str) -> String {
    let ns = Uuid::parse_str(NAMESPACE).expect("valid uuid namespace");
    Uuid::new_v5(&ns, summary.as_bytes()).to_string()
}
